src/core/lib/gprpp/per_cpu.h

```diff
@@ -51,23 +57,39 @@ class PerCpuOptions {
   size_t max_shards_ = std::numeric_limits<size_t>::max();
 };
 
+class PerCpuShardingHelper {
+ protected:
+  size_t GetShardingBits() {
+    if (GPR_UNLIKELY(state_.uses_until_refresh == 0)) state_ = State();
+    --state_.uses_until_refresh;
+    return state_.last_seen_cpu;
+  }
+
+ private:
+  struct State {
+    uint16_t last_seen_cpu = gpr_cpu_current_cpu();
+    uint16_t uses_until_refresh = 65535;
+  };
+  static thread_local State state_;
+};
+
 template <typename T>
-class PerCpu {
+class PerCpu : public PerCpuShardingHelper {
```

Comment:
nit: the "is a" relationship is unnecessary now. PerCpu can just own a helper and call `GetShardingBits()` since they share no state.